src/uwtools/template.py

```diff
@@ -0,0 +1,172 @@
+                            variable_to_substitute = v
+                else:
+                    more = re.search(pair.regex, var)
+                    if more is not None:
+                        new_value = cls.substitute_string(var, var_type, get_value)
+                        variable_to_substitute = variable_to_substitute.replace(var, new_value)
+        return variable_to_substitute
+
+    @classmethod
+    def substitute_structure(cls, structure_to_substitute, var_type: str, get_value):
+        """
+            Traverses a dictionary and substitutes variables in fields, lists
+            and nested dictionaries.
+        """
+        if isinstance(structure_to_substitute, dict):
+            for key, item in structure_to_substitute.items():
+                structure_to_substitute[key] = cls.substitute_structure(item, var_type, get_value)
+        elif isinstance(structure_to_substitute, Sequence) and not isinstance(structure_to_substitute, str):
+            for i, item in enumerate(structure_to_substitute):
+                structure_to_substitute[i] = cls.substitute_structure(item, var_type, get_value)
+        else:
+            structure_to_substitute = cls.substitute_string(structure_to_substitute, var_type,
+                                                            get_value)
+        return structure_to_substitute
+
+    @classmethod
+    def substitute_structure_from_environment(cls, structure_to_substitute):
+        return cls.substitute_structure(structure_to_substitute, TemplateConstants.DOLLAR_CURLY_BRACE, os.environ.get)
+
+    @classmethod
+    def substitute_with_dependencies(cls, dictionary, keys, var_type: str, shallow_precedence=True, excluded=()):
+        """
+            Given a dictionary with a complex (deep) structure, we want to substitute variables,
+            using keys, another dictionary that may also have a deep structure (dictionary and keys
+            can be the same dictionary if you want to substitute in place).
+            We create an index based on keys (see build_index) and substitute values in dictionary
+            using index. If variables may refer to other variables, more than one pass of substitution
+            may be needed, so we substitute until there is no more change in dictionary (convergence).
+        """
+        all_variables = cls.build_index(keys, excluded, shallow_precedence)
+        previous = {}
+        while dictionary != previous:
+            previous = copy.deepcopy(dictionary)
+            dictionary = cls.substitute_structure(dictionary, var_type, all_variables.get)
+        return dictionary
+
+    @classmethod
+    def build_index(cls, dictionary, excluded=None, shallow_precedence=True):
+        """
+            Builds an index of all keys with their values, going deep into the dictionary. The index
+            if a flat structure (dictionary).
+            If the same key name is present more than once in the structure, we want to
```

Comment:
What is the value in allowing the user/developer to specifying clobbering priority based on node depth when reading a configuration file?  How would a user/developer know where the values they want to be priority lay in the tree?  Is there a case where this is needed?